nixos/wpa_supplicant: add safe secret handling

rnhmjoj opened the pull request:
###### Motivation for this change

Possible solution of #87829.

###### Things done

- [x] Tested variable interpolation works
- [x] Tested the permissions of `/run/wpa_supplicant/wpa_supplicant.conf` are safe
- [x] Add release note
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

rnhmjoj committed:
nixos/wpa_supplicant: add safe secret handling

turion commented:
Does `wpa_supplicant` allow several config files? That would be an easier way than `sed`ing the passwords into play. Usually such solutions break at some point, e.g. if someone has an `@` in the password or so.

rnhmjoj commented:
> Does wpa_supplicant allow several config files?

No, I don't think it does. It does support multiple configuration files, but one file per network interface only.

> e.g. if someone has an @ in the password or so.

I have been very careful with that: the expansion happens if and only if the file contains `@VARIABLE@` where VARIABLE is an existing environment  variable. So, a single `@`  won't cause and expansion, similarly `myp@ssw@rd` won't be expanded, unless the user defined a `ssw` variable.

ncfavier commented:
But the sed will break if the password contains `#`

rnhmjoj commented:
> But the sed will break if the password contains #

Right, the `s` command implies reserving a character. I'm not sure how to do otherwise with sed.

rnhmjoj commented:
Idea, using awk to access the environment:
```
awk '{
  for(varname in ENVIRON)
    gsub("@"varname"@", ENVIRON[varname])
  print
}' "${configFile}" > "${finalConfig}"
```

ncfavier commented:
See https://mywiki.wooledge.org/TemplateFiles for more options

rnhmjoj commented:
I think using awk is strictly better: it does not expand missing variables, doesn't require perl and does not suffer from quoting issues.

rnhmjoj committed:
nixos/tests/wpa_supplicant: init

rnhmjoj commented:
I added a test with a few cases and it seems to be working.

ncfavier commented on nixos/modules/services/networking/wpa_supplicant.nix:
```diff
@@ -68,14 +74,27 @@ let
       wantedBy = [ "multi-user.target" ];
       stopIfChanged = false;
 
-      path = [ package ];
+      path = [ package pkgs.gawk ];
+      serviceConfig.RuntimeDirectory = "wpa_supplicant";
+      serviceConfig.EnvironmentFile = mkIf (cfg.environmentFile != null)
+        (builtins.toPath cfg.environmentFile);
 
       script =
       ''
-        if [ -f /etc/wpa_supplicant.conf -a "/etc/wpa_supplicant.conf" != "${configFile}" ]; then
+        if [ -f /etc/wpa_supplicant.conf -a "${toString configIsGenerated}" = 1 ]; then
           echo >&2 "<3>/etc/wpa_supplicant.conf present but ignored. Generated ${configFile} is used instead."
         fi
```
It's probably better to wrap the `if` in an `optionalString configIsGenerated`.

rnhmjoj replied:
Done.

ncfavier reviewed (commented):
It's not the most elegant thing ever but I don't think `wpa_supplicant` allows us to do anything cleverer, so... LGTM

ncfavier commented on nixos/modules/services/networking/wpa_supplicant.nix:
```diff
@@ -68,14 +74,27 @@ let
+      path = [ package pkgs.gawk ];
+      serviceConfig.RuntimeDirectory = "wpa_supplicant";
+      serviceConfig.EnvironmentFile = mkIf (cfg.environmentFile != null)
+        (builtins.toPath cfg.environmentFile);
```
`toPath` is [deprecated](https://nixos.org/manual/nix/stable/#builtin-toPath)

ncfavier replied:
Actually, why are you using `toPath` at all?

rnhmjoj replied:
Good question: I think I mean to write `toString`, in case `environmentFile` is a path, to avoid pulling it into the Nix store.

rnhmjoj commented:
I extended the test by making the daemon connect to an access point.

erikarvstedt reviewed (changes requested):
The test fails at assertion `assert "Connection established." in status`.

Build command: `nix build github:rnhmjoj/nixpkgs/e6ef8072f46aa299e8fef7589402821726cd06f2#nixosTests.wpa_supplicant`
Path: `/nix/store/14j348lhcgd5iddg9ck2adil9zzqk09h-vm-test-run-wpa_supplicant`

The complete output of `"echo status | wpa_cli -i wlan1` is:
```
wpa_cli v2.9
Copyright (c) 2004-2019, Jouni Malinen <j@w1.fi> and contributors

This software may be distributed under the terms of the BSD license.
See README for more details.



Interactive mode

> status
bssid=02:00:00:00:00:00
freq=2442
ssid=nixos-test
id=0
mode=station
pairwise_cipher=TKIP
group_cipher=TKIP
key_mgmt=WPA2-PSK
wpa_state=COMPLETED
p2p_device_address=42:00:00:00:01:00
address=02:00:00:00:01:00
uuid=572cf82f-c957-5653-9b16-b5cfb298abf1
```

ncfavier commented:
@ofborg test wpa_supplicant

erikarvstedt commented:
This might be a timing-related bug that's dependent on hardware. Maybe `Connection established.` is only shown in an intermediate state.

erikarvstedt commented:
This test derivations of this branch differ slightly depending on the path of the nixpkgs used for evaluation.
This was fixed in 9ce8df127d6d0b21ec3fc3864625677bb2fa73f6 which can be pulled in via a rebase to master.

doronbehar commented on nixos/modules/services/networking/wpa_supplicant.nix:
```diff
@@ -155,6 +176,19 @@ in {
+          the wireless configuration.
+          This option can be used to safely include secrets (PSKs, passwords,
+          etc.) in the <option>networking.wireless.networks</option> option via
+          environment variables.
```
```suggestion
          environment variables, with a <literal>@PASSWORD@</literal> as a value
          for <literal>psk</literal> or <literal>pskRaw</literal>.
```

doronbehar commented on nixos/modules/services/networking/wpa_supplicant.nix:
```diff
@@ -165,10 +199,14 @@ in {
-                in plaintext!
+                <warning><para>
+                  Be aware that this will be written to the nix store
+                  in plaintext! Use an enviroment variable instead.
```
```suggestion
                  in plaintext! Use an environment variable instead.
```

doronbehar commented on nixos/modules/services/networking/wpa_supplicant.nix:
```diff
@@ -179,7 +217,14 @@ in {
-                Mutually exclusive with <varname>psk</varname>.
+                <warning><para>
+                  Be aware that this will be written to the nix store
+                  in plaintext! Use an enviroment variable instead.
```
```suggestion
                  in plaintext! Use an environment variable instead.
```

doronbehar commented on nixos/modules/services/networking/wpa_supplicant.nix:
```diff
@@ -242,7 +287,15 @@ in {
-                Mutually exclusive with <varname>psk</varname> and <varname>pskRaw</varname>.
+                <warning><para>
+                  Be aware that this will be written to the nix store
+                  in plaintext! Use an enviroment variable for secrets.
```
```suggestion
                  in plaintext! Use an environment variable for secrets.
```

doronbehar reviewed (commented):
I don't use wpa_supplicant like this, so this is as much as I can help.

rnhmjoj committed:
nixos/release-notes: document wpa_supplicant changes

rnhmjoj commented:
@doronbehar Thank you. I have fixed the typos and added a small example to show how the variable are used.
I'm also planning on updating the NixOS manual about this change, but I will probably do it in another PR.

@erictapen I noticed the test is flaky, too. I'm looking into this, now.

rnhmjoj commented:
Should be fixed now.

rnhmjoj commented:
@GrahamcOfBorg test wpa_supplicant

erikarvstedt commented:
Here's are [some fixups](https://github.com/erikarvstedt/nixpkgs/commits/wpa-safe-ea).
`git fetch https://github.com/erikarvstedt/nixpkgs wpa-safe-ea:wpa-safe-ea`

rnhmjoj commented:
@erikarvstedt Thank you. I applied some of the changes which I consider to be an improvement, others are pretty subjective.

- 2ca16d39: I'm more familiar with the inverted syntax and it makes sense to me, given umask applies a "mask".
- 793935cb: Not sure what is gained by not putting awk in the path, at least the `preStart` script is cleaner without a long store path in it.
- 99148c43: I prefer the current style, it's also used in 58 other NixOS tests (`grep -zrlP '\w =\n +\'\'' nixos/tests/ | wc -l`).

erikarvstedt commented on nixos/modules/services/networking/wpa_supplicant.nix:
```diff
@@ -155,6 +176,26 @@ in {
+
+          For example, if you define your home wifi password as
+          <literal>PSK_HOME=mypassword</literal>, you can then refer to it with
+          the special syntax <literal>@MY_PASSWORD@</literal> in the value of
```
```suggestion
          the special syntax <literal>@PSK_HOME@</literal> in the value of
```

erikarvstedt reviewed (changes requested):
> I'm more familiar with the inverted syntax...

That's fine, too.

> Not sure what is gained by not putting awk in the path

It's a principle of Nix to reduce global state to a minimum. Putting `awk` in PATH implies that the service itself might require `awk` to operate when it's in fact used only at a single place in `script`.

>  I prefer the current style, ...

`<bikeshedding>`
But in **all** other places where you've introduced multiline strings you're using the formatting style that I suggested. 😉
`</bikeshedding>`

doronbehar reviewed (approved):
Diff LGTM overall, having an additional approval would be nice.

rnhmjoj commented:
> But in all other places where you've introduced multiline strings you're using the formatting style that I suggested. 😉

Yeah, I'm not consistent. I think I use the K&R style and the GNU style depending on the context.

erikarvstedt reviewed (approved):
Nice work!

turion commented on nixos/modules/services/networking/wpa_supplicant.nix:
```diff
@@ -155,6 +176,26 @@ in {
+          This option can be used to safely include secrets (PSKs, passwords,
+          etc.) in the <option>networking.wireless.networks</option> option.
+
+          For example, if you define your home wifi password as
```
Can you make this into an `example = ''[...]` and give a literal example of a valid file?

rnhmjoj replied:
I thought about this, but how would I show the file content given the option takes a filepath?
I would have to write `pkgs.writeText "secrets" ''...''`, which seems to suggest putting the secrets into the Nix store, again.

turion replied:
Ah true, so maybe give as example an example path to such a file, and then say in the description something along the line of "the file `/foo/bar.secrets` should contain contents `@MYPASS@=secret`"?

turion replied:
Or should it contain `MYPASS=secret`..? Well, an example would clarify that :)

rnhmjoj replied:
I does say `PSK_HOME=mypassword`. Anyway, I will write a manual chapter with a more detailed explanation, including why this variable dance is needed.

turion replied:
> I does say PSK_HOME=mypassword. 

Yes, I'm nitpicking here, what I find unclear is that this is what you need to write in the _file_. (And not e.g. in your config.)

ncfavier commented:
I think a cleaner alternative to exposing the `EnvironmentFile` to the user would be to add a `networking.wireless.networks.<name>.pskFile` option. If `pskFile` is set, write something like `psk=@psk_${ssid}@` to the config, and substitute that with the contents of the `pskFile` in the startup script.

rnhmjoj commented:
If you have many networks, which is pretty common, managing tens of `pskFile`s is not very handy.
Also, `psk` is not the only setting where one may use a secret: there's `pskRaw` and `auth` for WPA enterpise network. The latter is free-form, so there's no way to know where to expand the `pskFile`.

We could add a `pskFile` and `pskRawFile`, but they can't replace `environmentFile` entirely.

ncfavier commented:
OK, fair

rnhmjoj commented:
@turion I looked at similar options in NixOS and found a very good format for the example. What do you think, now?

erikarvstedt commented:
[Fixup](https://github.com/erikarvstedt/nixpkgs/commits/wpa-safe-ea-3) which simplifies the introductory text to the example. I'm close to the subject matter, but I still had troubles parsing the sentence 😉.

rnhmjoj commented:
Uhm, I think it's less clear: it doesn't explicitely say the variable are placeholders for secrets.
Let's wait for @turion to comment on this.

turion commented:
I find this very clear and helpful now. Thanks!

rnhmjoj commented:
@erikarvstedt I've amended the description with your last patch.
@turion Thank you!

rnhmjoj commented:
I think I've received enough feedback and the PR is in pretty good shape. Thank you all for helping out!
I'm merging this now, if something comes up it can be address it another PR.

rnhmjoj commented:
Note to myself: I should replace the current system with [this](https://w1.fi/cgit/hostap/commit/?id=e680a51e94a33591f61edb210926bcb71217a21a) once wpa_supplicant will make a new release.
